core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala

```diff
@@ -0,0 +1,267 @@
+      lazy val eventDeserializer = new CassandraJournal.EventDeserializer(serialization)
+      implicit def ec = materializer.executionContext
+
+      var seqNr = 0L // initialized in preStart
+      var partition = 0L
+      var count = 0L
+
+      var queryState: QueryState = QueryIdle
+
+      val newResultSetCb = getAsyncCallback[Try[ResultSet]] {
+        case Success(rs) =>
+          val (switchPartition, fetchMore) = queryState match {
+            case QueryInProgress(switchPartition, fetchMore) => (switchPartition, fetchMore)
+            case _ => throw new IllegalStateException(s"New ResultSet when in unexpected state $queryState")
+          }
+          val empty = rs.isExhausted() && !fetchMore
+          queryState = QueryResult(rs, empty, switchPartition)
+          tryPushOne()
+        case Failure(e) => failStage(e)
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case Continue =>
+          queryState match {
+            case QueryIdle          => query(switchPartition = false)
+            case _: QueryResult     => tryPushOne()
+            case _: QueryInProgress => // result will come
+          }
+      }
+
+      override def preStart(): Unit = {
+        system // fail fast if not ActorMaterializer
+
+        def partitionNr(sequenceNr: Long): Long =
+          (sequenceNr - 1L) / config.targetPartitionSize
+
+        queryState = QueryInProgress(switchPartition = false, fetchMore = false)
+        session.highestDeletedSequenceNumber(persistenceId).onComplete {
+          getAsyncCallback[Try[Long]] {
+            case Success(delSeqNr) =>
+              // lowest possible seqNr is 1
+              seqNr = math.max(delSeqNr + 1, math.max(fromSeqNr, 1))
+              // initial query
+              queryState = QueryIdle
+              query(switchPartition = false)
+
+            case Failure(e) => failStage(e)
+
+          }.invoke
+        }
+
+        refreshInterval match {
+          case Some(interval) =>
+            val initial =
+              if (interval >= 2.seconds) ThreadLocalRandom.current().nextLong(interval.toMillis).millis
+              else interval
+            schedulePeriodicallyWithInitialDelay(Continue, initial, interval)
+          case None =>
+        }
+      }
+
+      override def postStop(): Unit = {
+        // for GC, in case stage is still referenced for some reason
+        queryState = QueryIdle
+      }
+
+      def query(switchPartition: Boolean): Unit = {
+        queryState match {
+          case QueryIdle          => // good
+          case _: QueryInProgress => throw new IllegalStateException("Query already in progress")
+          case QueryResult(rs, _, _) =>
+            if (!rs.isExhausted()) throw new IllegalStateException("Previous query was not exhausted")
+        }
+        val pnr = if (switchPartition) partition + 1 else partition
+        queryState = QueryInProgress(switchPartition, fetchMore = false)
```

Comment:
do we need to reset switchPartition if we've incremented it?

Reply:
I don't think that is needed, or I don't see the scenario you were thinking of.

              // When ResultSet is exhausted we immediately look in next partition for more events.
              // We keep track of if the query was such switching partition and if result is empty
              // we complete the stage or wait until next Continue tick.